### PR TITLE
python311Packages.scikit-bio: unbreak, refactor, run tests, adopt

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

Last version bump broke the `scikit-bio` package, some dependencies were missing. This PR adds the missing dependency (`bio-format`), and refactors a lot of the code of the `scikit-bio` package.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/scikit-bio/default.nix [hunk 1]
```diff
@@ -1,70 +1,75 @@
...
     hdmedians
-    scikit-learn
+    biom-format
   ];
 
-  # cython package not included for tests
-  doCheck = false;
+  nativeCheckInputs = [
+    pytestCheckHook
+    coverage
```
Comment (natsukium) [inline pkgs/development/python-modules/scikit-bio/default.nix:57]:
```suggestion
```
Could you please remove `coverage` and patch if necessary?
We don't need to check the coverage.

File: pkgs/development/python-modules/biom-format/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    mkdir biom_tests
+    mv biom/tests biom_tests/tests
+    rm -r biom
+  '';
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pytestFlagsArray = [ "biom_tests/tests" ];
+
+  disabledTests = [ "test_to_dataframe_is_sparse" ];
```
Comment (natsukium) [inline pkgs/development/python-modules/biom-format/default.nix:52]: Could you explain the reason for disabling the test?

File: pkgs/development/python-modules/biom-format/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  src = fetchFromGitHub {
+    owner = "biocore";
+    repo = "biom-format";
+    rev = "refs/tags/${version}";
+    hash = "sha256-WRBc+C/UWme7wYogy4gH4KTIdIqU3KmBm2jWzGNxGQg=";
+  };
+
+  build-system = [
+    setuptools
+    cython
```
Comment (natsukium) [inline pkgs/development/python-modules/biom-format/default.nix:29]:
```suggestion
    cython
    numpy
```
numpy is required not only at runtime but also at build time.
https://github.com/biocore/biom-format/blob/2.1.15/pyproject.toml

File: pkgs/development/python-modules/scikit-bio/default.nix [hunk 4]
```diff
@@ -1,53 +1,75 @@
...
 }:
 
 buildPythonPackage rec {
-  version = "0.6.0";
-  format = "setuptools";
   pname = "scikit-bio";
+  version = "0.6.0";
+  pyproject = true;
+
   disabled = !isPy3k;
```
Comment (natsukium) [inline pkgs/development/python-modules/scikit-bio/default.nix:28]:
```suggestion
```
We don't care about Python2 anymore.

File: pkgs/development/python-modules/scikit-bio/default.nix [hunk 5]
```diff
@@ -1,53 +1,75 @@
...
     homepage = "http://scikit-bio.org/";
     description = "Data structures, algorithms and educational resources for bioinformatics";
-    license = licenses.bsd3;
-    platforms = [ "x86_64-linux" "x86_64-darwin" ];
-    maintainers = [ ];
+    license = lib.licenses.bsd3;
+    platforms = [
+      "x86_64-linux"
+      "x86_64-darwin"
+    ];
```
Comment (natsukium) [inline pkgs/development/python-modules/scikit-bio/default.nix:72]:
```suggestion
```
I think we can remove this restriction. It seems to be able to build on aarch64.

File: pkgs/development/python-modules/biom-format/default.nix (hunk 2 above)
Comment (TomaSajt) [reply pkgs/development/python-modules/biom-format/default.nix]: Upstream already has a fix for it, fetched it it fetchpatch

File: pkgs/development/python-modules/scikit-bio/default.nix (hunk 1 above)
Comment (TomaSajt) [reply pkgs/development/python-modules/scikit-bio/default.nix]: Removed, no issues

File: pkgs/development/python-modules/biom-format/default.nix (hunk 3 above)
Comment (TomaSajt) [reply pkgs/development/python-modules/biom-format/default.nix:39]: Thanks

**Review by natsukium — APPROVED** [review]
LGTM. I'm waiting for the CI.

**Pull request merged**